### PR TITLE
Warn if quota limits for a company are reached

### PR DESCRIPTION
**Pull request by fjvallarino**
## Description
Checks if the company has gone above 50% or 80% of their available quota and logs a message in both cases.

Some changes were required to surface the necessary quota headers.

## Motivation and Context
It's part of our epic.

## How Has This Been Tested?
Several unit tests were added and modified. The existing endpoints still work correctly, for example: https://services-stage.risevision.com/twitter/get-tweets?companyId=121ae986-1b89-40c8-b10b-a64c60fb9e03&username=CNN&count=10

## Release Plan:
- As the Submitter, upon requesting review of this pull request, I confirm that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed. 
- As the Reviewer, upon approving the changes in this PR, I confirm I have reviewed and I agree that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed

#### Release Checklist Items Skipped?
No.

@santiagonoguez @stulees please review

### CODE REVIEW

File: src/timelines/index.js [hunk 1]
```diff
@@ -13,8 +13,11 @@ const {
 } = constants;
 
 const validationErrorFor = message => Promise.reject(new Error(message));
+const quotaLimitError = {message: "Quota limit reached."};
 
 const currentTimestamp = () => new Date().getTime();
+const quotaNormalPct = 0.5;
+const quotaSeverePct = 0.2;
```
Comment (santiagonoguez) [inline src/timelines/index.js]: @fjvallarino I think these should me moved to config.js

File: src/timelines/index.js [hunk 2]
```diff
@@ -77,6 +80,40 @@ const saveStatus = (query) => {
...
+const checkUserQuota = (query) => {
+  const {companyId} = query;
+
+  return cache.getUserQuotaFor(companyId)
+  .then(quota => {
+    if (!quota || quota.remaining > 0 || quota.resetTs < currentTimestamp() / SECONDS) {
+      return Promise.resolve();
+    }
+
+    const err = {...quota, quotaLimitReached: true};
```
Comment (santiagonoguez) [inline src/timelines/index.js]: @fjvallarino yes, this way is much better, thanks !

File: src/timelines/index.js [hunk 3]
```diff
@@ -77,6 +80,40 @@ const saveStatus = (query) => {
...
+  const remaining = apiRes.status === CONFLICT_ERROR ? 0 : apiRes.quota && apiRes.quota.remaining;
+  const resetTs = apiRes.quota && apiRes.quota.resetTs;
+  const quota = {remaining, resetTs};
+
+  return apiRes.quota && apiRes.quota.valid ? cache.saveUserQuota(companyId, quota) : Promise.resolve();
+}
+
+const logUserQuota = (companyId, quota) => {
+  if (!quota || !quota.total || !quota.remaining) {
+    console.warn(`Missing rate limit headers for company: ${companyId}`);
```
Comment (santiagonoguez) [inline src/timelines/index.js]: @fjvallarino curious, is there some case where you have seen this happening ?

File: src/timelines/index.js [hunk 4]
```diff
@@ -77,6 +80,40 @@ const saveStatus = (query) => {
...
+  } else if (quota.remaining < quota.total * quotaSeverePct) {
+    console.warn(`Current quota usage above 80% for company: ${companyId}`);
+  } else if (quota.remaining < quota.total * quotaNormalPct) {
+    console.warn(`Current quota usage above 50% for company: ${companyId}`);
```
Comment (santiagonoguez) [inline src/timelines/index.js]: @fjvallarino maybe as a future improvement we could derive these hardcoded 80% and 50% from quotaSeverPct and quotaNormalPct variables, thoughts ?

File: test/unit/timelines.tests.js [hunk 5]
```diff
@@ -266,190 +278,342 @@ describe("Timelines", () => {
...
         assert(res.json.called);
+        assert(!res.status.called);
+      });
+    });
 
-        const data = res.json.lastCall.args[0];
-        assert(data);
-        assert(data.tweets);
-        assert.equal(25, data.tweets.length);
+    it("should resolve if remaining quota is greather than 0", () => {
```
Comment (santiagonoguez) [inline test/unit/timelines.tests.js]: @fjvallarino typo   greather -> greater

File: test/unit/timelines.tests.js [hunk 6]
```diff
@@ -266,190 +278,342 @@ describe("Timelines", () => {
...
+        tlResponse.quota.total = 100;
+        tlResponse.quota.resetTs = (Date.now() / SECONDS) + 100;
+
+        return timelines.handleGetTweetsRequest(req, res)
+        .then(() => {
+          assert(cache.getUserQuotaFor.called);
+
+          assert(console.warn.called);
+          assert.equal(console.warn.lastCall.args[0], "Current quota usage above 80% for company: test");
+        });
```
Comment (santiagonoguez) [inline test/unit/timelines.tests.js]: @fjvallarino this file is growing large now; I separated all cache related tests to another test file, and I think we should do the same for quota related tests, thoughts ?

File: src/timelines/index.js (hunk 4 above)
Comment (fjvallarino) [reply src/timelines/index.js]: @santiagonoguez updated as part of the move the config.js

File: src/timelines/index.js (hunk 3 above)
Comment (fjvallarino) [reply src/timelines/index.js]: @santiagonoguez not really, and the fact that it generates quite a bit of noise in the tests' console output made me think about removing it. Any preference?

File: test/unit/timelines.tests.js (hunk 5 above)
Comment (fjvallarino) [reply test/unit/timelines.tests.js]: @santiagonoguez updated!

File: src/timelines/index.js (hunk 3 above)
Comment (santiagonoguez) [reply src/timelines/index.js]: @fjvallarino I think it's safer to leave it for now.

File: src/timelines/index.js (hunk 1 above)
Comment (fjvallarino) [reply src/timelines/index.js]: @santiagonoguez updated!

**Pull request merged**